lib/devise/parameter_sanitizer.rb

```diff
@@ -1,99 +1,153 @@
+    #
+    # === Arguments
+    #
+    # * +operation+ - A +Symbol+ with the operation that the controller is
+    #   performing, like +sign_up+, +sign_in+, etc.
+    #
+    # === Examples
+    #
+    #    # Inside the `RegistrationsController#create` action.
+    #    resource = build_resource(devise_parameter_sanitizer.sanitize(:sign_up))
+    #    resource.save
+    #
+    # Returns an +ActiveSupport::HashWithIndifferentAccess+ with the permitted
+    # attributes.
+    def sanitize(operation)
+      if block = @blocks[operation]
+        cast_to_hash block.call(default_params)
       else
-        default_for(kind)
+        cast_to_hash default_sanitize(operation)
       end
     end
 
-    def sanitize(kind)
-      if block = @blocks[kind]
-        block.call(default_params)
+    def for(operation, &block)
+      if block_given?
+        @blocks[operation] = block
       else
-        default_sanitize(kind)
+        default_attributes_for(operation)
       end
     end
 
-    private
-
-    def default_for(kind)
-      raise ArgumentError, "a block is expected in Devise base sanitizer"
-    end
-
-    def default_sanitize(kind)
-      default_params
-    end
-
-    def default_params
-      params.fetch(resource_name, {})
-    end
-  end
-
-  class ParameterSanitizer < BaseSanitizer
-    def initialize(*)
-      super
-      @permitted = Hash.new { |h,k| h[k] = attributes_for(k) }
-    end
+    protected
 
+    # Shortcut to retrieve the sanitized attributes for the +sign_in+ operation.
+    # Returns an +ActiveSupport::HashWithIndifferentAccess+ with the permitted
+    # attributes for the Sign In operation.
     def sign_in
-      permit self.for(:sign_in)
+      permit default_attributes_for(:sign_in)
     end
 
+    # Returns an +ActiveSupport::HashWithIndifferentAccess+ with the permitted
+    # attributes for the Sign Up operation.
     def sign_up
-      permit self.for(:sign_up)
+      permit default_attributes_for(:sign_up)
     end
 
+    # Shortcut for the@p permitted attributes for the account update operation.
+    # Setting a specific list of attributes through the +for+ method will take
+    #
+    #
+    # Returns an +ActiveSupport::HashWithIndifferentAccess+ with the permitted
+    # attributes for the Account Update operation.
     def account_update
-      permit self.for(:account_update)
+      permit default_attributes_for(:account_update)
     end
 
     private
 
-    # TODO: We do need to flatten so it works with strong_parameters
-    # gem. We should drop it once we move to Rails 4 only support.
+    # Cast a sanitized +ActionController::Parameters+ to a +HashWithIndifferentAccess+
+    # that can be used elsewhere.
+    #
+    # Returns an +ActiveSupport::HashWithIndifferentAccess+.
+    def cast_to_hash(params)
+      if params
+        # Rails 4.1 doesn't have `to_unsafe_h`, only `to_hash`.
+        hash = params.respond_to?(:to_unsafe_h) ? params.to_unsafe_h : params.to_hash
```

Comment:
It looks like this is only called with parameters that have already been permitted - in which case just calling .to_h would be better